### ISSUE TITLE
Torch does not optimize matrix multiplication on certain transformer architectures

### ISSUE THREAD

**Issue by rumoschilling** [pytorch/pytorch#194473]
### 🐛 Describe the bug

If the attention layer is Attention @ Value @ Linear ... (the dots stand for potential other layers in the future), pytorch always computes Attention @ Value first. This is the correct choice most of the time, as it is convention to have input-dim < output-dim. But if this is not the case, Value @ Linear first is faster. The same thing applies for the fusing linear layer of different heads. There it seems more common to actually have this multiplication and not have some layer norm or residual connections in between, but on the other hand it is less likely that the head-out dim > linear-out.

I looked up that this is not done in the documentation and then let an LLM model test this. I am certainly less competent than LLMs at specific benchmarking methods, and I told it to test a lot of cases so the original code to test this benchmark is quite large. For completeness I asked it to provide a minimal reproducer, but as stated I cannot check its full correctness. 

Another objection the LLM provided was that numerically, matrix multiplication is not associative, but my counterargument is that reordering is not less stable so this should not be an argument for making the operation slower.

[minimal_attention_reassociation_repro.py](https://github.com/user-attachments/files/31340720/minimal_attention_reassociation_repro.py)

### Versions

Collecting environment information...
PyTorch version: 2.11.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Omarchy (x86_64)
GCC version: (GCC) 16.2.1 20260810
Clang version: 22.1.8
CMake version: Could not collect
Libc version: glibc-2.44

Python version: 3.14.7 (main, Aug 10 2026, 07:46:56) [GCC 16.1.1 20260728] (64-bit runtime)
Python platform: Linux-7.1.8-arch1-3-x86_64-with-glibc2.44
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           48 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  12
On-line CPU(s) list:                     0-11
Vendor ID:                               AuthenticAMD
Model name:                              AMD Ryzen 5 7640U w/ Radeon 760M Graphics
CPU family:                              25
Model:                                   116
Thread(s) per core:                      2
Core(s) per socket:                      6
Socket(s):                               1
Stepping:                                1
Microcode version:                       0xa70410a
Frequency boost:                         enabled
CPU(s) scaling MHz:                      68%
CPU max MHz:                             4972.4351
CPU min MHz:                             405.9140
BogoMIPS:                                6987.10
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpuid_fault cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid overflow_recov succor smca fsrm flush_l1d amd_lbr_pmc_freeze
Virtualization:                          AMD-V
L1d cache:                               192 KiB (6 instances)
L1i cache:                               192 KiB (6 instances)
L2 cache:                                6 MiB (6 instances)
L3 cache:                                16 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-11
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Mitigation; Clear CPU buffers
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.1.0.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.19.0.56
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.0
[pip3] nvidia-nccl-cu13==2.28.9
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.11.0
[pip3] triton==3.6.0
[conda] Could not collect

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by Ash20pk** [comment]
Took a shot at this in #194507 (draft). The approach is a post-grad Inductor pass rather than anything in eager: flatten a chain of `mm`/`bmm` nodes, run the standard matrix-chain-order DP over the shared dims, and only rewrite when the result is strictly cheaper in MACs than the order already in the graph. It handles chains longer than three and skips any intermediate with more than one consumer, since reassociating there would just force a recompute.

On your associativity point — I agree reordering isn't systematically *less* stable, but it isn't bit-identical either, so the pass is behind an off-by-default `config.reassociate_matmul` for now. Turning it on by default seems like it should be its own change backed by a real model suite.

Fair warning on the numbers: I don't have a GPU, so the only measurement in the PR is CPU (1.36 ms -> 0.19 ms on a 1024x1024x1024x8 chain). The FLOP model is launch- and bandwidth-blind, which is exactly where I'd expect a reviewer to push back for CUDA.

Drafted with an AI coding agent; I reviewed the code and the reasoning above before posting.

**Comment by rumoschilling** [comment]
Did you run it on other test cases? As I said before my knowledge of the library is limited, but I would be wary of the  attention being recognized in the compute graph and not masked by some abstraction, especially for multi head being some concatenation in between. Sorry if this is all caught by the compute graph, like I said I am usually quite an outsider to the low level things, but it seems like testing it on a basic matmul chain is not sufficient to catch dome pathological behavior.

**Comment by Ash20pk** [comment]
@rumoschilling Added an attention test and you're right, it doesn't fire. After decomposition nn.Linear on a 3D input becomes reshape → mm → reshape, the attention product is a bmm while the projections are mm, and with SDPA the score/value product is one fused op — so the pass never sees an adjacent chain. Pinned that in the tests as a known gap (test_attention_output_projection_is_not_reassociated_yet). As it stands the PR only helps literal mm/bmm chains; reaching attn @ V @ W_o means looking through the head-merge views, which I'll look at separately.

**Comment by rumoschilling** [comment]
Hey @Ash20pk thank you so much for the work you already put in. I am just checking in whether you are still looking into the remaining issues or whether another compiler wizard can tackle it. Cheers!

**Comment by Ash20pk** [comment]
Hey @rumoschilling I will pick this up this weekend, thanks for checking in
